docs/topics/multiplatform/multiplatform-discover-project.md

```diff
@@ -1,169 +1,315 @@
+    * If the code is used for only one target, it should be defined in a specific source set for that target (for
+      example, `jvmMain` for the JVM).
+* Code written in platform-specific source sets can access declarations from the common source set. For example, the
+  code in `jvmMain` can use code from `commonMain`. However, the opposite isn't true: `commonMain` can't use code
+  from `jvmMain`.
+* Code written in platform-specific source sets can use the corresponding platform dependencies. For example, the code
+  in `jvmMain` can use Java-only libraries, like [Guava](https://github.com/google/guava)
+  or [Spring](https://spring.io/).
+
+### Intermediate source sets
+
+In simple multiplatform projects, only common code is shared among all available platforms and platform-specific code.
+In practice, you often need more granular code sharing.
+
+Consider an example where you need to target all modern Apple devices and Android devices:
+
+```kotlin
+kotlin {
+    android()
+    iosArm64()         // 64-bit iPhone devices
+    macosArm64()   // Modern Apple Silicon-based Macs
+    watchosX64()    // Modern 64-bit Apple Watch devices
+    tvosArm64()       // Modern Apple TV devices  
 }
 ```
 
-</tab>
-</tabs>
+And you need a source set to add a function that generates a UUID for all Apple devices:
+
+```kotlin
+import platform.Foundation.NSUUID
+
+fun randomUuidString(): String {
+    // You want to access Apple-specific APIs
+    return NSUUID().UUIDString()
+}
+```
+
+You can't add this function to `commonMain`. `commonMain` is compiled to all declared targets, including Android,
+but `platform.Foundation.NSUUID` is an Apple-specific API that's not available on Android. Kotlin shows an error if you
+try to reference it in `commonMain`.
+
+You could copy and paste this code to each Apple-specific source
+set: `iosArm64Main`, `macosArm64Main`, `watchosX64Main`, and `tvosArm64Main`. But this approach is not recommended
+because duplicating code like this is prone to errors.
+
+To solve this issue, you can use _intermediate source sets_. An intermediate source set is a Kotlin source set that
+compiles to some, but not all of the targets in the project. You can also see intermediate source sets referred to as
+hierarchical source sets or simply hierarchies.
+
+Kotlin creates some intermediate source sets by default. In this specific case, the resulting project structure will
+look like this:
+
+![Intermediate source sets](intermediate-source-sets-diagram.svg){width=700}
+
+Here, the gray blocks at the bottom are platform-specific source sets. Target labels are omitted for clarity.
+
+The `appleMain` block is an intermediate source set created by Kotlin for sharing code compiled to Apple-specific
+targets. The `appleMain` source set compiles to only Apple targets. Kotlin allows using Apple-specific APIs
+in `appleMain`, and you can add the `randomUUID()` function here.
```

Comment:
> The `appleMain` source set compiles to only Apple targets. Kotlin allows using Apple-specific APIs in `appleMain`, and you can add the `randomUUID()` function here.

The causal link between these two sentences is important. We do want people to understand that Kotlin looks at the set of compiled targets when it decided which APIs are OK to use, and I've explicitly tried to repeat that consistently over the article. 

Suggestion:

> The `appleMain` source set compiles to only Apple targets. Therefore, Kotlin allows using Apple-specific APIs in `appleMain`, and you can add the `randomUUID()` function here.